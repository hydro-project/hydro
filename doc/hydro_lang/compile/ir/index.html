<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `ir` mod in crate `hydro_lang`."><title>hydro_lang::compile::ir - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="hydro_lang" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (f6092f224 2025-09-22)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module ir</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../hydro_lang/index.html">hydro_<wbr>lang</a><span class="version">0.14.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module ir</a></h2><h3><a href="#modules">Module Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="../index.html">In hydro_<wbr>lang::<wbr>compile</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../../index.html">hydro_lang</a>::<wbr><a href="../index.html">compile</a></div><h1>Module <span>ir</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../../src/hydro_lang/compile/ir/mod.rs.html#1-3162">Source</a> </span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><dl class="item-table"><dt><a class="mod" href="backtrace/index.html" title="mod hydro_lang::compile::ir::backtrace">backtrace</a></dt><dd>Platform-independent interface for collecting backtraces, used in the Hydro IR to
trace the origin of each node.</dd></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.DebugExpr.html" title="struct hydro_lang::compile::ir::DebugExpr">Debug<wbr>Expr</a></dt><dd>Wrapper that displays only the tokens of a parsed expr.</dd><dt><a class="struct" href="struct.DebugInstantiateFinalized.html" title="struct hydro_lang::compile::ir::DebugInstantiateFinalized">Debug<wbr>Instantiate<wbr>Finalized</a></dt><dt><a class="struct" href="struct.DebugType.html" title="struct hydro_lang::compile::ir::DebugType">Debug<wbr>Type</a></dt><dd>Debug displays the type’s tokens.</dd><dt><a class="struct" href="struct.HydroIrMetadata.html" title="struct hydro_lang::compile::ir::HydroIrMetadata">Hydro<wbr>IrMetadata</a></dt><dt><a class="struct" href="struct.HydroIrOpMetadata.html" title="struct hydro_lang::compile::ir::HydroIrOpMetadata">Hydro<wbr>IrOp<wbr>Metadata</a></dt><dd>Metadata that is specific to the operator itself, rather than its outputs.
This is available on <em>both</em> inner nodes and roots.</dd><dt><a class="struct" href="struct.QMacroSimplifier.html" title="struct hydro_lang::compile::ir::QMacroSimplifier">QMacro<wbr>Simplifier</a></dt><dd>AST visitor that simplifies q! macro expansions</dd><dt><a class="struct" href="struct.TeeNode.html" title="struct hydro_lang::compile::ir::TeeNode">TeeNode</a></dt></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.BuildersOrCallback.html" title="enum hydro_lang::compile::ir::BuildersOrCallback">Builders<wbr>OrCallback</a></dt><dt><a class="enum" href="enum.DebugInstantiate.html" title="enum hydro_lang::compile::ir::DebugInstantiate">Debug<wbr>Instantiate</a></dt><dt><a class="enum" href="enum.HydroNode.html" title="enum hydro_lang::compile::ir::HydroNode">Hydro<wbr>Node</a></dt><dd>An intermediate node in a Hydro graph, which consumes data
from upstream nodes and emits data to downstream nodes.</dd><dt><a class="enum" href="enum.HydroRoot.html" title="enum hydro_lang::compile::ir::HydroRoot">Hydro<wbr>Root</a></dt><dd>An root in a Hydro graph, which is an pipeline that doesn’t emit
any downstream values. Traversals over the dataflow graph and
generating DFIR IR start from roots.</dd><dt><a class="enum" href="enum.HydroSource.html" title="enum hydro_lang::compile::ir::HydroSource">Hydro<wbr>Source</a></dt><dd>A source in a Hydro graph, where data enters the graph.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.dbg_dedup_tee.html" title="fn hydro_lang::compile::ir::dbg_dedup_tee">dbg_<wbr>dedup_<wbr>tee</a></dt><dt><a class="fn" href="fn.deep_clone.html" title="fn hydro_lang::compile::ir::deep_clone">deep_<wbr>clone</a></dt><dt><a class="fn" href="fn.emit.html" title="fn hydro_lang::compile::ir::emit">emit</a></dt><dt><a class="fn" href="fn.transform_bottom_up.html" title="fn hydro_lang::compile::ir::transform_bottom_up">transform_<wbr>bottom_<wbr>up</a></dt><dt><a class="fn" href="fn.traverse_dfir.html" title="fn hydro_lang::compile::ir::traverse_dfir">traverse_<wbr>dfir</a></dt></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><dl class="item-table"><dt><a class="type" href="type.SeenTeeLocations.html" title="type hydro_lang::compile::ir::SeenTeeLocations">Seen<wbr>TeeLocations</a></dt><dt><a class="type" href="type.SeenTees.html" title="type hydro_lang::compile::ir::SeenTees">Seen<wbr>Tees</a></dt></dl></section></div></main></body></html>