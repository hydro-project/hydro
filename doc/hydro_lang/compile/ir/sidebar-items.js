window.SIDEBAR_ITEMS = {"enum":["BoundKind","BuildersOrCallback","CollectionKind","DebugInstantiate","HydroNode","HydroRoot","HydroSource","KeyedSingletonBoundKind","StreamOrder","StreamRetry"],"fn":["dbg_dedup_tee","deep_clone","emit","transform_bottom_up","traverse_dfir"],"mod":["backtrace"],"struct":["DebugExpr","DebugInstantiateFinalized","DebugType","HydroIrMetadata","HydroIrOpMetadata","QMacroSimplifier","TeeNode"],"trait":["DfirBuilder"],"type":["SeenTeeLocations","SeenTees"]};