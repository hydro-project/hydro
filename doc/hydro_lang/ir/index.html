<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `ir` mod in crate `hydro_lang`."><title>hydro_lang::ir - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="hydro_lang" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (10fa3c449 2025-04-26)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../hydro_lang/index.html">hydro_<wbr>lang</a><span class="version">0.13.2</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module ir</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#enums" title="Enums">Enums</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#types" title="Type Aliases">Type Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate hydro_<wbr>lang</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">hydro_lang</a></div><h1>Module <span>ir</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/hydro_lang/ir.rs.html#1-2433">Source</a> </span></div><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.DebugExpr.html" title="struct hydro_lang::ir::DebugExpr">Debug<wbr>Expr</a></dt><dd>Debug displays the type’s tokens.</dd><dt><a class="struct" href="struct.DebugInstantiateFinalized.html" title="struct hydro_lang::ir::DebugInstantiateFinalized">Debug<wbr>Instantiate<wbr>Finalized</a></dt><dt><a class="struct" href="struct.DebugType.html" title="struct hydro_lang::ir::DebugType">Debug<wbr>Type</a></dt><dd>Debug displays the type’s tokens.</dd><dt><a class="struct" href="struct.HydroIrMetadata.html" title="struct hydro_lang::ir::HydroIrMetadata">Hydro<wbr>IrMetadata</a></dt><dt><a class="struct" href="struct.TeeNode.html" title="struct hydro_lang::ir::TeeNode">TeeNode</a></dt></dl><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><dl class="item-table"><dt><a class="enum" href="enum.BuildersOrCallback.html" title="enum hydro_lang::ir::BuildersOrCallback">Builders<wbr>OrCallback</a></dt><dt><a class="enum" href="enum.DebugInstantiate.html" title="enum hydro_lang::ir::DebugInstantiate">Debug<wbr>Instantiate</a></dt><dt><a class="enum" href="enum.HydroLeaf.html" title="enum hydro_lang::ir::HydroLeaf">Hydro<wbr>Leaf</a></dt><dd>An leaf in a Hydro graph, which is an pipeline that doesn’t emit
any downstream values. Traversals over the dataflow graph and
generating DFIR IR start from leaves.</dd><dt><a class="enum" href="enum.HydroNode.html" title="enum hydro_lang::ir::HydroNode">Hydro<wbr>Node</a></dt><dd>An intermediate node in a Hydro graph, which consumes data
from upstream nodes and emits data to downstream nodes.</dd><dt><a class="enum" href="enum.HydroSource.html" title="enum hydro_lang::ir::HydroSource">Hydro<wbr>Source</a></dt><dd>A source in a Hydro graph, where data enters the graph.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.dbg_dedup_tee.html" title="fn hydro_lang::ir::dbg_dedup_tee">dbg_<wbr>dedup_<wbr>tee</a></dt><dt><a class="fn" href="fn.deep_clone.html" title="fn hydro_lang::ir::deep_clone">deep_<wbr>clone</a></dt><dt><a class="fn" href="fn.emit.html" title="fn hydro_lang::ir::emit">emit</a></dt><dt><a class="fn" href="fn.transform_bottom_up.html" title="fn hydro_lang::ir::transform_bottom_up">transform_<wbr>bottom_<wbr>up</a></dt><dt><a class="fn" href="fn.traverse_dfir.html" title="fn hydro_lang::ir::traverse_dfir">traverse_<wbr>dfir</a></dt></dl><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><dl class="item-table"><dt><a class="type" href="type.SeenTeeLocations.html" title="type hydro_lang::ir::SeenTeeLocations">Seen<wbr>TeeLocations</a></dt><dt><a class="type" href="type.SeenTees.html" title="type hydro_lang::ir::SeenTees">Seen<wbr>Tees</a></dt></dl></section></div></main></body></html>