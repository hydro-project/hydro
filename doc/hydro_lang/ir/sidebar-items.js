window.SIDEBAR_ITEMS = {"enum":["DebugInstantiate","HydroLeaf","HydroNode","HydroSource"],"fn":["dbg_dedup_tee"],"struct":["DebugExpr","TeeNode"],"type":["SeenTees"]};