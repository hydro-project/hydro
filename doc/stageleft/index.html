<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `stageleft` crate."><title>stageleft - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-fecee23daa4e8a6d.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="stageleft" data-themes="" data-resource-suffix="" data-rustdoc-version="1.83.0-nightly (4ac7bcbaa 2024-09-04)" data-channel="nightly" data-search-js="search-a99f1315e7cc5121.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-921df33f47b8780c.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../stageleft/index.html">stageleft</a><span class="version">0.4.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#attributes">Attribute Macros</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">stageleft</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/stageleft/lib.rs.html#1-355">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="internal/index.html" title="mod stageleft::internal">internal</a></div></li><li><div class="item-name"><a class="mod" href="runtime_support/index.html" title="mod stageleft::runtime_support">runtime_<wbr>support</a></div></li></ul><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.PATH_SEPARATOR.html" title="macro stageleft::PATH_SEPARATOR">PATH_<wbr>SEPARATOR</a></div></li><li><div class="item-name"><a class="macro" href="macro.q.html" title="macro stageleft::q">q</a></div><div class="desc docblock-short">Creates a quoted expression for Hydroflow+.</div></li><li><div class="item-name"><a class="macro" href="macro.quse_fn.html" title="macro stageleft::quse_fn">quse_fn</a></div></li><li><div class="item-name"><a class="macro" href="macro.stageleft_crate.html" title="macro stageleft::stageleft_crate">stageleft_<wbr>crate</a></div></li><li><div class="item-name"><a class="macro" href="macro.stageleft_no_entry_crate.html" title="macro stageleft::stageleft_no_entry_crate">stageleft_<wbr>no_<wbr>entry_<wbr>crate</a></div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.BorrowBounds.html" title="struct stageleft::BorrowBounds">Borrow<wbr>Bounds</a></div></li><li><div class="item-name"><a class="struct" href="struct.RuntimeData.html" title="struct stageleft::RuntimeData">Runtime<wbr>Data</a></div><div class="desc docblock-short">Represents a piece of data that will be passed into the generated code</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.IntoQuotedMut.html" title="trait stageleft::IntoQuotedMut">Into<wbr>Quoted<wbr>Mut</a></div></li><li><div class="item-name"><a class="trait" href="trait.IntoQuotedOnce.html" title="trait stageleft::IntoQuotedOnce">Into<wbr>Quoted<wbr>Once</a></div></li><li><div class="item-name"><a class="trait" href="trait.Quoted.html" title="trait stageleft::Quoted">Quoted</a></div></li><li><div class="item-name"><a class="trait" href="trait.QuotedContext.html" title="trait stageleft::QuotedContext">Quoted<wbr>Context</a></div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.quote_type.html" title="fn stageleft::quote_type">quote_<wbr>type</a></div><div class="desc docblock-short">Captures a fully qualified path to a given type, which is useful when
the generated code needs to explicitly refer to a type known at staging time.</div></li></ul><h2 id="attributes" class="section-header">Attribute Macros<a href="#attributes" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="attr" href="attr.entry.html" title="attr stageleft::entry">entry</a></div><div class="desc docblock-short">Defines an entrypoint for staged code, which will be available as a proc macro.
The entrypoint must be a function that returns <code>impl Quoted&lt;T&gt;</code> for some type <code>T</code>.</div></li><li><div class="item-name"><a class="attr" href="attr.runtime.html" title="attr stageleft::runtime">runtime</a></div><div class="desc docblock-short">Marks a chunk of code as being runtime-only, which means that no staged code in its crate can depend on it.
Code behind this attribute is allowed to use staged entrypoints defined in the same crate.</div></li><li><div class="item-name"><a class="attr" href="attr.top_level_mod.html" title="attr stageleft::top_level_mod">top_<wbr>level_<wbr>mod</a></div><div class="desc docblock-short">A utility for declaring top-level public modules in a Stageleft crate that exports macros.</div></li></ul></section></div></main></body></html>