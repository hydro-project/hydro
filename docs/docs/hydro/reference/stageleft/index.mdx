---
title: Stageleft
sidebar_position: 0
---

Hydro uses a library called Stageleft to power the `q!` macro and code generation logic. 

For detailed documentation on Stageleft's architecture and the `q!` macro, see the [Stageleft README](https://github.com/hydro-project/stageleft/blob/main/README.md).

:::note

Hydro programmers need not understand how Stageleft works under the hood; this information is provided here for readers interested in how its `q!` macro works.

:::
