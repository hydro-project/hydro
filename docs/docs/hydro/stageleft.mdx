---
title: Stageleft
sidebar_position: 10
---

import StageleftDocs from 'stageleft/README.md'

Under the hood, Hydro uses a library called Stageleft to power the `q!` macro and code generation logic. The following docs, from the Stageleft README, outline the core architecture of Stageleft.

:::note

Hydro programmers need not understand how to use Stageleft; this information is provided here for readers
interested in how its `q!` macro works.

:::

<StageleftDocs/>
