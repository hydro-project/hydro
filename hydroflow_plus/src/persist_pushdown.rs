use crate::ir::*;

fn predicate_pushdown_node(node: HfPlusNode) -> HfPlusNode {
    match node {
        HfPlusNode::Map {
            f,
            input: box HfPlusNode::Persist(behind_persist),
        } => HfPlusNode::Persist(Box::new(predicate_pushdown_node(HfPlusNode::Map {
            f,
            input: behind_persist,
        }))),
        o => o,
    }
}

fn predicate_pushdown_leaf(leaf: HfPlusLeaf) -> HfPlusLeaf {
    match leaf {
        HfPlusLeaf::ForEach { f, input } => {
            let input = predicate_pushdown_node(*input);
            HfPlusLeaf::ForEach {
                f,
                input: Box::new(input),
            }
        }
        o => o,
    }
}

pub fn predicate_pushdown(ir: Vec<HfPlusLeaf>) -> Vec<HfPlusLeaf> {
    ir.into_iter().map(predicate_pushdown_leaf).collect()
}

#[stageleft::runtime]
#[cfg(test)]
mod tests {
    use stageleft::*;

    use crate::{Location, MultiGraph};

    #[test]
    fn predicate_pushdown_through_map() {
        let flow = crate::builder::FlowBuilder::<MultiGraph>::new();
        let process = flow.process(&());

        process
            .source_iter(q!(0..10))
            .all_ticks()
            .map(q!(|v| v + 1))
            .for_each(q!(|n| println!("{}", n)));

        let built = flow.build();

        insta::assert_debug_snapshot!(&built.ir);

        let pushed_down = built.optimize_with(super::predicate_pushdown);

        insta::assert_debug_snapshot!(&pushed_down.ir);
    }
}
