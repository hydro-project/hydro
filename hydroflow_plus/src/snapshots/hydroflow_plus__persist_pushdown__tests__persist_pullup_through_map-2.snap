---
source: hydroflow_plus/src/persist_pushdown.rs
expression: "&optimized.ir"
---
[
    ForEach {
        f: { use crate :: __staged :: persist_pushdown :: tests :: * ; | n | println ! ("{}" , n) },
        input: Persist(
            Map {
                f: { use crate :: __staged :: persist_pushdown :: tests :: * ; | v | v + 1 },
                input: Source {
                    source: Iter(
                        { use crate :: __staged :: persist_pushdown :: tests :: * ; 0 .. 10 },
                    ),
                    produces_delta: false,
                    location_id: 0,
                },
            },
        ),
    },
]
